src/containers/habit-tracker/analytics/year-view/Heatmap.tsx

```diff
@@ -25,192 +25,246 @@ const MONTH_NAMES = [
+        const month = MONTH_NAMES[monthIndex - 1];
+
+        // Get the log count
+        const logs = series[seriesIndex][dataPointIndex];
+
+        // Custom tooltip display
+        return `<div style="padding: 5px; color: ${colors.textGrey}">
+                  <div style="color: ${colors.blue}; font-weight: bold;">${month} ${dayOfMonth}
+                  </div>
+                  Logs: ${logs}
+                </div>`;
+      },
     },
-  },
-});
+  };
+};
 
+// Interface for the Heatmap component
 interface HeatmapProps {
-  data: HeatmapEntry[];
+  data: { name: string; data: { x: number; y: number; day: number }[] }[];
 }
 
-// Functional component to render the heatmap - used in the year view of the analytics page of the habit tracker
+// Functional component to render the heatmap chart
 export function Heatmap({ data }: HeatmapProps) {
-  const series = useMemo(() => {
-    // Create a map to group logs by the day of the week
-    const weekDayToLogsMap = new Map<
-      string,
-      { x: number; y: number; day: number }[]
-    >();
-
-    // Iterate over the data and populate the map
-    data.forEach(({ month, dayOfWeek, logCount, dayOfMonth }) => {
-      // Get existing logs for the day of the week, or initialize a new array
-      const logs = weekDayToLogsMap.get(dayOfWeek.toString()) || [];
-
-      // Add the current log entry to the array
-      logs.push({ x: month, y: logCount, day: dayOfMonth });
-
-      // Update the map with the new log entry
-      weekDayToLogsMap.set(dayOfWeek.toString(), logs);
-    });
-
-    // Convert the map to an array of objects that is suitable for ApexCharts
-    return Array.from(weekDayToLogsMap.entries()).map(([weekDay, logs]) => ({
-      name: weekDay,
-      data: logs,
-    }));
+  const [loading, setLoading] = useState(true);
+  const [rendered, setRendered] = useState(false);
+  const [windowWidth, setWindowWidth] = useState(window.innerWidth);
+
+  // Event listener to update windowWidth on resize
+  useEffect(() => {
+    const handleResize = () => {
+      setWindowWidth(window.innerWidth);
+    };
+
+    window.addEventListener("resize", handleResize);
+
+    return () => {
+      window.removeEventListener("resize", handleResize);
+    };
+  }, []);
+
+  const handleRendered = () => {
+    console.log("Chart rendered, setting 'rendered' to true");
+    setRendered(true);
+  };
+
+  const options = useHeatmapOptions(windowWidth, handleRendered);
+
+  useEffect(() => {
+    if (data.length > 0) {
+      console.log("Data loaded, setting 'loading' to false");
+      setLoading(false);
+    }
   }, [data]);
 
-  const options = HeatmapOptions();
+  // Fallback: Remove spinner after 1.5 seconds if mounted/animationEnd doesn't fire
+  useEffect(() => {
+    if (!rendered) {
+      const timeout = setTimeout(() => {
+        console.log("Fallback: forcing spinner to disappear after 1.5 seconds");
+        setRendered(true);
+      }, 1500);
```

Comment:
I do not like this. What if the connection is slow and loading takes longer? We need to find a better solution here where we remove the spinner only after we know the component has rendered.
I don't know the exact solution yet. I'll have to play around with it to figure it out. I'll let you know once I do. :)